requests_auth/authentication.py

```diff
@@ -186,25 +186,51 @@ def __init__(self, token_url: str, username: str, password: str, **kwargs):
             self.data["scope"] = " ".join(scope) if isinstance(scope, list) else scope
         self.data.update(kwargs)
 
+        # As described in https://tools.ietf.org/html/rfc6749#section-6
+        self.refresh_data = {
+            "grant_type": "refresh_token",
+            "refresh_token": None  # the refresh token is set when a refresh is imminent
+        }
+        if scope:
+            self.refresh_data["scope"] = " ".join(scope) if isinstance(scope, list) else scope
+        self.refresh_data.update(kwargs)
+
         all_parameters_in_url = _add_parameters(self.token_url, self.data)
         self.state = sha512(all_parameters_in_url.encode("unicode_escape")).hexdigest()
 
     def __call__(self, r):
-        token = OAuth2.token_cache.get_token(self.state, self.request_new_token)
+        token = OAuth2.token_cache.get_token(
+            key=self.state,
+            on_missing_token=self.request_new_token,
+            on_refresh_token=self.refresh_token,
+        )
         r.headers[self.header_name] = self.header_value.format(token=token)
         return r
 
     def request_new_token(self):
         # As described in https://tools.ietf.org/html/rfc6749#section-4.3.3
-        token, expires_in = request_new_grant_with_post(
+        token, expires_in, refresh_token = request_new_grant_with_post(
             self.token_url,
             self.data,
             self.token_field_name,
             self.timeout,
             self.session,
         )
         # Handle both Access and Bearer tokens
-        return (self.state, token, expires_in) if expires_in else (self.state, token)
+        return (self.state, token, expires_in, refresh_token) if expires_in else (self.state, token)
+
+    def refresh_token(self, refresh_token: str):
+        # As described in https://tools.ietf.org/html/rfc6749#section-6
+        self.refresh_data['refresh_token'] = refresh_token
+        token, expires_in, refresh_token = request_new_grant_with_post(
+            self.token_url,
+            self.refresh_data,
+            self.token_field_name,
+            self.timeout,
+            self.session
+        )
+        # Handle both Access and Bearer tokens
+        return (self.state, token, expires_in, refresh_token) if expires_in else (self.state, token)
```

Comment:
When requesting a refresh, is it possible not to get an expiry? this condition make sense when requesting a token, as it can be an access token or a bearer token, however can we still encounter this case for a refresh?

Reply:
The specification only defines refreshing for access tokens. https://tools.ietf.org/html/rfc6749#section-6
If the refresh call is successful, an access token is issued: "If valid and authorized, the authorization server issues an access token as described in Section 5.1."

However, the `expires_in` field is not mandatory according to the specification. It is just recommended for access tokens. https://tools.ietf.org/html/rfc6749#section-5.1

Reply:
I removed the Bearer part as the response of the refresh call is per definition an access token.